nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-service-aws/src/main/java/org/apache/nifi/kafka/service/aws/AmazonMSKConnectionService.java

```diff
@@ -69,6 +155,172 @@ protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final String stsEndpoint = propertyContext.getProperty(AWS_WEB_IDENTITY_STS_ENDPOINT).getValue();
+        final SSLContextProvider sslContextProvider = propertyContext.getProperty(AWS_WEB_IDENTITY_SSL_CONTEXT_SERVICE).asControllerService(SSLContextProvider.class);
+
+        final ApacheHttpClient.Builder httpClientBuilder = ApacheHttpClient.builder();
+
+        if (sslContextProvider != null) {
+            final SSLContext sslContext = sslContextProvider.createContext();
+            httpClientBuilder.socketFactory(new SSLConnectionSocketFactory(sslContext));
+        }
+
+        final StsClientBuilder stsClientBuilder = StsClient.builder().httpClient(httpClientBuilder.build());
+
+        if (!StringUtils.isBlank(stsRegionId)) {
+            stsClientBuilder.region(Region.of(stsRegionId));
+        }
+
+        if (!StringUtils.isBlank(stsEndpoint)) {
+            stsClientBuilder.endpointOverride(URI.create(stsEndpoint));
+        }
+
+        final StsClient stsClient = stsClientBuilder.build();
+
+        return new WebIdentityCredentialsProvider(stsClient, tokenProvider, roleArn, roleSessionName, sessionSeconds);
+    }
+
+    private static final class WebIdentityCredentialsProvider implements AwsCredentialsProvider, AutoCloseable {
+        private static final Duration SKEW = Duration.ofSeconds(60);
+
+        private final StsClient stsClient;
+        private final OAuth2AccessTokenProvider tokenProvider;
+        private final String roleArn;
+        private final String roleSessionName;
+        private final Integer sessionSeconds;
+
+        private volatile AwsSessionCredentials cachedCredentials;
+        private volatile Instant expiration;
+
+        private WebIdentityCredentialsProvider(final StsClient stsClient,
+                                               final OAuth2AccessTokenProvider tokenProvider,
+                                               final String roleArn,
+                                               final String roleSessionName,
+                                               final Integer sessionSeconds) {
+            this.stsClient = Objects.requireNonNull(stsClient, "stsClient required");
+            this.tokenProvider = Objects.requireNonNull(tokenProvider, "tokenProvider required");
+            this.roleArn = Objects.requireNonNull(roleArn, "roleArn required");
+            this.roleSessionName = Objects.requireNonNull(roleSessionName, "roleSessionName required");
+            this.sessionSeconds = sessionSeconds;
+        }
+
+        @Override
+        public AwsCredentials resolveCredentials() {
+            final Instant now = Instant.now();
+            final AwsSessionCredentials currentCredentials = cachedCredentials;
+            final Instant currentExpiration = expiration;
+
+            if (currentCredentials != null && currentExpiration != null && now.isBefore(currentExpiration.minus(SKEW))) {
+                return currentCredentials;
```

Comment:
Recommend refactoring to a single return. It may also be useful to add another private method for evaluating the expiration that encapsulates null checking.